### ISSUE TITLE
Set Postgres schema as a separate field instead of connection string.

### ISSUE THREAD

**Issue by amit-reflectivity** [prisma/prisma#5337]
## Problem

The environment variable, DATABASE_URL, can't be changed and does not contain the schema. The environment variable can change over time and cannot make another environment variable based on DATABASE_URL.

## Suggested solution

Add a field called schema in a datasource definition like provider and url.

## Alternatives

1. Allow url to be defined as -> url = env("DATABASE_URL")env("SCHEMA) or url = env("DATABASE_URL") + "?schema=alternate"
2. Do not assume a schema of public by default. This would mean not prefix tables with public that would allow Postgres search_path to be used for matching tables.

## Additional context
Using a command like "prisma introspect" would benefit by the added flexibility outside of code.

**Comment by chdalski** [comment]
I would very much appreciate this feature (separation of concerns).

An easy solution could be something like:
`
@schema("Auth")
model User {
  id    Int    @id @default(autoincrement())
  name  String @db.VarChar(200)
  posts Post[]
}
`

But that would get you into trouble if you have more than one "User" model.

So ultimately you would need something like:
`
@path(schema="Auth", table="User")
model AuthUser {
  id    Int    @id @default(autoincrement())
  name  String @db.VarChar(200)
}
`

`
@path(schema="Guest", table="User")
model GuestUser {
  id    Int    @id @default(autoincrement())
  name  String @db.VarChar(200)
}
`

**Comment by janpio** [comment]
@chdalski What you describe would probably be part of https://github.com/prisma/prisma/issues/1122, vs. the issue above - which is a very technical detail of how you get the schema into your database URL.

For that we actually have a different solution already implemented: https://www.prisma.io/docs/guides/development-environment/environment-variables#expanding-variables

> The environment variable, DATABASE_URL, can't be changed and does not contain the schema. The environment variable can change over time and cannot make another environment variable based on DATABASE_URL.

So for this we would advise you to compose a new env variable `DATABASE_URL_WITH_SCHEMA` via your `.env` file based on the one that comes from your environment and can not be changed, and then use that in your schema:
```
DATABASE_URL_WITH_SCHEMA=${DATABASE_URL}?schema=foo
```

Does this work for you and your use case @amit-reflectivity?

**Comment by amit-reflectivity** [comment]
@janpio This does not work in my use case.

I can't make an environmental variable based on another environment variable.

**Comment by janpio** [comment]
Why not? Does the documentation I linked to not work as described, or are there other reasons from doing it like this?

**Comment by amit-reflectivity** [comment]
The environments other than development cannot make an environment variable from another environment variable
such as DATABASE_URL_WITH_SCHEMA=${DATABASE_URL}?schema=foo

The DATABASE_URL environment variable is set automatically and can change dynamically. When the change happens it automatically redeploys to make use of the change in value.

**Comment by chdalski** [comment]
> @chdalski What you describe would probably be part of #1122, vs. the issue above - which is a very technical detail of how you get the schema into your database URL.
@janpio Thank you :+1:

**Comment by janpio** [comment]
> The environments other than development cannot make an environment variable from another environment variable
> such as DATABASE_URL_WITH_SCHEMA=${DATABASE_URL}?schema=foo

Yes, that is exactly what is described in the documentation I linked to about how you can use a `.env` file to achieve this. Please read the documentation and try it out.

**Comment by amit-reflectivity** [comment]
I can't use .env file in staging and production. I am limited in the way of doing environment variables. It automatically assigns DATABSASE_URL and it dynamically changes. You cannot create an environment variable based on another environment variable.

I can have .env file in development environment only. Which I do, but I impose the limitation in development.

**Comment by janpio** [comment]
Even though there is no actual value in the `.env` file, but just logic to take an env var and stitch it together with another one to create a combined env var? Interesting. Is that a rule on your side (convention) or is that enforced somehow by some software?

This is indeed a new use case we did not consider yet.
